Ruler: remote rule evaluation hardening (#8785)
grafana/loki: docs/sources/operations/scalability.md

In compute-constrained environments, garbage collection can become a significant performance factor. Frequently-run garbage collection interferes with running the application by using CPU resources. The use of memory ballast can mitigate the issue. Memory ballast allocates extra, but unused virtual memory in order to inflate the quantity of live heap space. Garbage collection is triggered by the growth of heap space usage. The inflated quantity of heap space reduces the perceived growth, so garbage collection occurs less frequently.

Configure memory ballast using the ballast_bytes configuration option.

> ## Remote Rule Evaluation

Reviewer: ```suggestion
## Remote rule evaluation
```

> This feature was introduced as part of [`LID-0002`](https://github.com/grafana/loki/pull/8129).

Reviewer: I don't think this line is necessary.  
It is also incorrect, the LID is just the design doc and discussion, it doesn't change the code or introduce new features.

Author: You're right, I could've worded it better; I want to include the design doc so folks can read up on the design considerations if they want to.

How's this?

```suggestion
This feature was initial proposed in [`LID-0002`](https://github.com/grafana/loki/pull/8129).
```

> ### Background & Identification of Problem

Reviewer: Your overview text is fine, but I would remove both of the headings:
"Background & Identification of Problem" 
"Recommended Solution"
You're documenting a feature here, not a bug fix.  All of the paragraph text from lines 37 - 51 can sit right under the "Remote rule evaluation" heading.

Author: OK 👍

By default, the `ruler` component embeds a query engine to evaluate rules. This generally works fine, except when rules
are complex or have to process a large amount of data regularly. Poor performance of the `ruler` manifests as recording rules metrics
with gaps or missed alerts. This situation can be detected by alerting on the `cortex_prometheus_rule_group_iterations_missed_total` metric
when it has a non-zero value.

### Recommended Solution

> A solution to this problem is to externalize rule evaluation from the `ruler` process. The `ruler` embedded query engine

Reviewer 2: Do we have a guideline whether to have a newline between heading and paragraph? I see both variants.

Author: I don't think it matters; the markdown renderer will sort it out - but we could add a linter @jdbaldry?

is single-threaded, meaning that rules are not split, sharded, or otherwise accelerated like regular Loki queries. The `query-frontend`
component exists explicitly for this purpose and, when combined with a number of `querier` instances, can massively
improve rule evaluation performance and lead to fewer missed iterations.

It is generally recommended to create a separate `query-frontend` deployment and `querier` pool from your existing one - which handles adhoc
queries via Grafana, `logcli`, or the API. Rules should be given priority over adhoc queries because they are used to produce
metrics or alerts which may be crucial to the reliable operation of your service; if you use the same `query-frontend` and `querier` pool
for both, your rules will be executed with the same priority as adhoc queries which could lead to unpredictable performance.

### Implementation

> To enable remote rule evaluation, set the following configuration options:

Reviewer: Where do you set this configuration?  In which file?

Author: Loki only supports one configuration file; the other files are for overrides, and it would be clear from context that this belongs in the config file. I can add a link to the config docs page, if that'll make it more clear?

```yaml
ruler:
  evaluation:
    mode: remote
    query_frontend:
      address: dns:///<query-frontend-service>:<grpc-port>
```

> The `ruler` component will become simply a gRPC client to the `query-frontend` service (side-note: this will result in much lower

Reviewer: ```suggestion
When you enable remote rule evaluation, the `ruler` component becomes a gRPC client to the `query-frontend` service.  This may result in lower resource usage of your rulers.
```

Author: Agree with the change except `s/may/will` because the ruler is now doing significantly less work. I'll update it 👍

> resource usage of your `ruler`s!). Rules' LogQL queries will be executed against the given `query-frontend` service.

Reviewer: ```suggestion
The LogQL queries coming from the Ruler will be executed against the given `query-frontend` service.
```

Requests will be load-balanced across all `query-frontend` IPs if the `dns:///` prefix is used.

> Queries that fail to execute are _not_ retried at this time.

Reviewer: I think this is worth calling out in a note.
```suggestion
> **Note:** Queries that fail to execute are _not_ retried.
```

Author: Good call 👍
